java-security/Migration_SAPJavaBuildpackProjects.md

```diff
@@ -1,53 +1,239 @@
+
+**Please note, this Migration Guide is only intended for applications using SAP Java Buildpack.**
 You're using the SAP Java Buildpack, if you can find the `sap_java_buildpack` in the deployment descriptor of your application, e.g. in your `manifest.yml` file.
 
-## Maven Dependencies
-To use the latest SAP-internal `java-container-security` (version `3.2.6`) the dependencies declared in maven `pom.xml` need to be updated.
+## Overview of all breaking changes
+
+The following list is an overview of the breaking changes that might require changes in existing
+applications.
+
+- New security related maven dependencies, see [Migration section](#migrate).
+- `AccessToken` instead of `XSUserInfo` in tomcat principal object. See [User principal](#user-principal).
+- Principal name is in new format, see [Principal name](#principal-name).
 
-First make sure you have the following dependencies defined in your pom.xml:
+## Migrate an existing buildpack java app <a name="migrate"></a>
+
+Your buildpack application has probably a dependency to one of the old security api projects in its `pom.xml`:
 
 ```xml
 <dependency>
-  <groupId>com.sap.cloud.security.xsuaa</groupId>
-  <artifactId>api</artifactId>
-  <version>2.6.0</version>
-  <scope>provided</scope>
+    <groupId>com.sap.cloud.security.xsuaa</groupId>
+    <artifactId>java-container-security-api</artifactId>
+    <scope>provided</scope>
 </dependency>
 ```
 
-Now you are ready to **remove** the **`java-container-security`** client library by deleting the following lines from the pom.xml:
+Or
+
 ```xml
 <dependency>
-  <groupId>com.sap.xs2.security</groupId>
-  <artifactId>java-container-security</artifactId>
+    <groupId>com.sap.cloud.security.xsuaa</groupId>
+    <artifactId>api</artifactId>
+    <scope>provided</scope>
 </dependency>
+```
+
+> If you do not have any of the api projects above as dependency you can skip this chapter!
+
+Those must be replaced with the new api project:
+
+```xml
 <dependency>
-  <groupId>com.sap.xs2.security</groupId>
-  <artifactId>java-container-security-api</artifactId>
+    <groupId>com.sap.cloud.security</groupId>
+    <artifactId>java-api</artifactId>
+    <version>2.6.0</version>
+    <scope>provided</scope>
 </dependency>
 ```
-Or
+
+> Scope can be set to provided because it is already shipped with the buildpack.
+> Take a look at the [version overview](https://wiki.wdf.sap.corp/wiki/display/xs2java/Component+Versions)
+> to see which version of the api is provided with the buildpack.
+
+Now you will most likely have compile error at places where the old api module was used. Those are the parts of the
+application where information from the token (or user info) was read. See the next sections on how to migrate that.
+
+## User principal <a name="user-principal"></a>
+
+In the old version of the buildpack the user principal could be obtained via:
+
+```java
+XSUserInfo userInfo = (XSUserInfo) request.getUserPrincipal();
```

Comment:
Is this possible again? With revert of UserInfoHolder...?

Reply:
No, not in the current state of the integration. The principal does not implement both interfaces (`XSUserInfo` and `AccessToken`).
What is possible now is that the user loads the `UserInfo` from the `UserInfoHolder`.

Reply:
okay, but then there is no benefit from using the XSUserInfoAdapter...

Reply:
We need the adapter to support XSUserHolder (otherwise would we need the old library).
Additionally the adapter enables an easier transition for the users.